# XML, HTML, and Web Scraping

JSON and XML are two different ways to represent hierarchical data. Which one is better? There are lots of articles online which discuss similarities and differences between JSON and XML and their advantages and disadvantages. Both formats are still in current usage, so it is good to be familiar with both. However, JSON is more common, so we'll focus on working with JSON representations of hierarchical data.

The reading covered an example of using Beautiful Soup to parse XML. Rather than doing another example XML now, we'll skip straight to scraping HTML from a webpage. Both HTML and XML can be parsed in a similar way with Beautiful Soup.

In [1]:
import pandas as pd
import requests as rq
from bs4 import BeautifulSoup

## Scraping an HTML table with Beautiful Soup

Open the URL https://en.wikipedia.org/wiki/List_of_United_States_cities_by_population and scroll down until you see a table of the cities in the U.S. with population over 100,000 (as of Jul 1, 2022). We'll use Beautiful Soup to scrape information from this table.

Read in the HTML from the ULR using the `requests` library.

In [2]:
response = rq.get("https://en.wikipedia.org/wiki/List_of_United_States_cities_by_population")

Use Beautiful Soup to parse this string into a tree called `soup`

In [3]:
soup = BeautifulSoup(response.content, "html.parser")

To find an HTML tag corresponding to a specific element on a webpage, right-click on it and choose "Inspect element". Go to the cities table Wikipedia page and do this now.

You should find that the cities table on the Wikipedia page corresponds to the element

```
<table class="wikitable sortable jquery-tablesorter" style="text-align:center">
```

In [4]:
# in this case
# <table class="sortable wikitable sticky-header-multi static-row-numbers sort-under col1left col2center jquery-tablesorter" style="text-align:right">

There are many `<table>` tags on the page.

In [5]:
len(soup.find_all("table"))

10

In [6]:
# print(soup.find_all("table"))

We can use attributes like `class=` and `style=` to narrow down the list.

In [7]:
len(soup.find_all("table",
                  attrs={
                      "class": "sortable wikitable sticky-header-multi static-row-numbers sort-under col1left col2center",
                      "style": "text-align:right"}
                  ))

1

At this point, you can manually inspect the tables on the webpage to find that the one we want is the first one (see `[0]` below). We'll store this as `table`.

In [8]:
table = soup.find_all("table",
                  attrs={
                      "class": "sortable wikitable sticky-header-multi static-row-numbers sort-under col1left col2center",
                      "style": "text-align:right"}
                  )

In [9]:
# print(table)

**Now you will write code to scrape the information in `table` to create a Pandas data frame with one row for each city and columns for: city, state, population (2022 estimate), and 2020 land area (sq mi).** Refer to the Notes/suggestions below as you write your code. A few Hints are provided further down, but try coding first before looking at the hints.

Notes/suggestions:

- Use as a guide the code from the reading that produced the data frame of Statistics faculty
- Inspect the page source as you write your code
- You will need to write a loop to get the information for all cities, but you might want to try just scraping the info for New York first
- You will need to pull the text from the tag. If `.text` returns text with "\n" at the end, try `.get_text(strip = True)` instead of `.text`
- Don't forget to convert to a Pandas Data Frame; it should have 333 rows and 4 columns
- The goal of this exercise is just to create the Data Frame. If you were going to use it --- e.g., what is the population density for all cities in CA? --- then you would need to clean the data first (to clean strings and convert to quantitative). (You can use Beautiful Soup to do some of the cleaning for you, but that goes beyond our scope.)

In [10]:
# initialize an empty list
rows = []

table_tag = table[0]

# iterate over all rows in the cities table
for cities in table_tag.find_all("tr")[1:]:

    # Get all the cells (<td>) in the row.
    cells = cities.find_all("td")

    if not cells: ## fix suggested by GeminiAI - error explainer
      continue

    # The information we need is the text between tags.

    # Find the city in cell[0]
    # which is contained in the <a> tag
    city_tag = cells[0].find("a") or cells[0]
    city = city_tag.text

    # Find the state of the city in cell[1]
    # which for most states is contained in the <a> tag
    state_tag = cells[1].find("a") or cells[1]
    state = state_tag.text

    # Find the population of the city in cell[2]
    # which for most faculty is contained in the <td> tag
    pop_tag = cells[2]
    pop = pop_tag.text

    # Find the land area of the city in cell[5]
    # which for most faculty is contained in the <td> tag
    area_tag = cells[5]
    land_area = area_tag.text

    # Append this data.
    rows.append({
        "City": city,
        "State": state,
        "Population (estimate 2023)": pop,
        "Land area (square miles)": land_area
    })

In [11]:
# rows

In [12]:
df = pd.DataFrame(rows)
# print(df)

In [13]:
# Remove the \n characters from 'Population (estimate 2023)' and 'Land area (square miles)' ## assisted by ChatGPT
df['Population (estimate 2023)'] = df['Population (estimate 2023)'].str.replace('\n', '').str.replace(',', '').str.strip().astype(int)
df['Land area (square miles)'] = df['Land area (square miles)'].str.replace('\n', '').str.replace(',', '').str.strip().astype(float)

print(df)

            City State  Population (estimate 2023)  Land area (square miles)
0       New York    NY                     8258035                     300.5
1    Los Angeles    CA                     3820914                     469.5
2        Chicago    IL                     2664452                     227.7
3        Houston    TX                     2314157                     640.4
4        Phoenix    AZ                     1650070                     518.0
..           ...   ...                         ...                       ...
331         Yuma    AZ                      100858                     120.7
332  New Bedford    MA                      100695                      20.0
333      Suffolk    VA                      100659                     399.2
334     Hesperia    CA                      100633                      72.7
335    Davenport    IA                      100354                      63.8

[336 rows x 4 columns]
            City State  Population (estimate 2023)  

Hints:

- Each city is a row in the table; find all the `<tr>` tags to find all the cities
- Look for the `<td>` tag to see table entries within a row
- The rank column is represented by `<th>` tags, rather than `<td>` tags. So within a row, the first (that is, `[0]`) `<td>` tag corresponds to the city name.

## Aside: Scraping an HTML table with Pandas



The Pandas command `read_html` can be used to scrape information from an HTML table on a webpage.

We can call `read_html` on the URL.

In [14]:
pd.read_html("https://en.wikipedia.org/wiki/List_of_United_States_cities_by_population")

URLError: <urlopen error [SSL: CERTIFICATE_VERIFY_FAILED] certificate verify failed: unable to get local issuer certificate (_ssl.c:1000)>

However, this scrapes all the tables on the webpage, not just the one we want. As with Beautiful Soup, we can narrow the search by specifying the table attributes.

In [ ]:
pd.read_html("https://en.wikipedia.org/wiki/List_of_United_States_cities_by_population", attrs = {"class": "sortable wikitable sticky-header-multi static-row-numbers sort-under col1left col2center", "style": "text-align:right"})

This still returns 3 tables. As we remarked above, the table that we want is the first one (see `[0]` below).

In [15]:
df_cities2 = pd.read_html("https://en.wikipedia.org/wiki/List_of_United_States_cities_by_population", attrs = {"class": "sortable wikitable sticky-header-multi static-row-numbers sort-under col1left col2center", "style": "text-align:right"})[0]
df_cities2

URLError: <urlopen error [SSL: CERTIFICATE_VERIFY_FAILED] certificate verify failed: unable to get local issuer certificate (_ssl.c:1000)>

Wait, that seemed much easier than using Beautiful Soup, and it returned a data frame, and we even got for free some formatting like removing the commas from the population! Why didn't we just use `read_html` in the first place? It's true the `read_html` works well when scraping information from an HTML *table*. Unfortunately, you often want to scrape information from a webpage that isn't conveniently stored in an HTML table, in which case `read_html` won't work. (It only searches for `<table>`, `<th>`, `<tr>`, and `<td>` tags, but there are many other HTML tags.) Though Beautiful Soup is not as simple as `read_html`, it is more flexible and thus more widely applicable.

## Scraping information that is NOT in a `<table>` with Beautiful Soup

The Cal Poly course catalog http://catalog.calpoly.edu/collegesandprograms/collegeofsciencemathematics/statistics/#courseinventory contains a list of courses offered by the Statistics department. **You will scrape this website to obtain a Pandas data frame with one row for each DATA or STAT course and two columns: course name and number (e.g, DATA 301. Introduction to Data Science) and term typically offered (e.g., Term Typically Offered: F, W, SP).**

Note: Pandas `read_html` is not help here since the courses are not stored in a `<table>.`

In [ ]:
pd.read_html("http://catalog.calpoly.edu/collegesandprograms/collegeofsciencemathematics/statistics/#courseinventory")


Notes/suggestions:


- Inspect the page source as you write your code
- The courses are not stored in a `<table>`. How are they stored?
- You will need to write a loop to get the information for all courses, but you might want to try just scraping the info for DATA 100 first
- What kind of tag is the course name stored in? What is the `class` of the tag?
- What kind of tag is the quarter(s) the course is offered stored in? What is the `class` of the tag? Is this the only tag of this type with the class? How will you get the one you want?
- You don't have to remove the number of units (e.g., 4 units) from the course name and number, but you can try it if you want
- You will need to pull the text from the tag. If `.text` returns text with "\n" at the end, try `get_text(strip = True)` instead of `text`
- Don't forget to convert to a Pandas Data Frame; it should have 74 rows and 2 columns
- The goal of this exercise is just to create the Data Frame. If you were going to use it then you might need to clean the data first. (You can use Beautiful Soup to do some of the cleaning for you, but that goes beyond our scope.)



In [16]:
import requests
response = requests.get("https://catalog.calpoly.edu/collegesandprograms/collegeofsciencemathematics/statistics/#courseinventory")

In [17]:
from bs4 import BeautifulSoup
soup = BeautifulSoup(response.content, "html.parser")

In [18]:
# soup

In [19]:
results = soup.find_all("div", {"class": "courseblock"})

In [20]:
# results

In [21]:
# lists to store data
course_names = []
terms_offered = []

# loop to parse each courseblock result # Assisted by ChatGPT

for block in results:

  # finding course name
  course_name_tag = block.find("p", {"class": "courseblocktitle"})
  # extracting text with span # assissted by GeminiAI
  if course_name_tag.find("span"):
    course_name_tag.find("span").extract()
  course_name = course_name_tag.get_text(strip=True)

  # finding term offered
  term_offered = block.find("p", {"class": "noindent"}).get_text(strip=True).replace("Term Typically Offered: ", '')

  # appending to lists
  course_names.append(course_name)
  terms_offered.append(term_offered)

# creating dataframe

df = pd.DataFrame({"Course Name": course_names, "Term Typically Offered": terms_offered})

print(df)

                                          Course Name Term Typically Offered
0                   DATA 100. Data Science for All I.               F, W, SP
1             DATA 301. Introduction to Data Science.               F, W, SP
2          DATA 401. Data Science Process and Ethics.                      F
3   DATA 402. Mathematical Foundations of Data Sci...                      F
4         DATA 403. Data Science Projects Laboratory.                      F
..                                                ...                    ...
69             STAT 551. Statistical Learning with R.                      F
70           STAT 566. Graduate Consulting Practicum.                     SP
71                STAT 570. Selected Advanced Topics.                    TBD
72          STAT 590. Graduate Seminar in Statistics.               F, W, SP
73                                  STAT 599. Thesis.                    TBD

[74 rows x 2 columns]
                                          Course Name

Hints:

- Each course is represented by a `<div>` with `class=courseblock`, so you can find all the courses with `soup.find_all("div", {"class": "courseblock"})`
- The course name is in a `<p>` tag with `class=courseblocktitle`, inside a `<strong>` tag. (Though I don't think we need to find the strong tag here.)
- The term typically offered is in `<p>` tag with `class=noindent`. However, there are several tags with this class; term typically offered is the first one.
- If you want to use Beautiful Soup to remove the course units (e.g., 4 units), find the `<span>` tag within the course name tag and `.extract()` this span tag